### DISCUSSION TITLE
[Feedback Requested] Cross-platform Adaptation for Bottom Navigation Bars (Tab Bars)

### DISCUSSION THREAD

**Discussion by leighajarett** [flutter/uxr#94] [Cross-platform Design]
# What is a navigation bar?
A navigation bar offers a persistent and convenient way to switch between primary destinations in an app. It is pinned to the bottom of the screen, and usually provides access to 3-5 destinations. Each destination is represented by an icon and optional text label.

# Platform differences
On the latest version of Android, app bars follow the design guidelines defined in [Material 3](https://m3.material.io/components/navigation-bar/overview). On iOS, an equivalent component is defined in [Apple’s Human Interface Guidelines](https://developer.apple.com/design/human-interface-guidelines/components/navigation-and-search/tab-bars/) called tab bars.

|| Navigation Bar in M3  |  Tab bar in HIG |
|---|---|---|
|| <img alt="Navigation Bar in M3 " src="https://user-images.githubusercontent.com/21065911/231519349-145dd3bd-7230-4f36-a45a-8ced55283870.png">  |  <img alt="ab bar in HIG" src="https://user-images.githubusercontent.com/21065911/231519370-0108fe1a-ce95-4d00-ae79-eb221d42301a.png"> |
| Container height | 80 (108 with pill-shaped device navigation, higher on devices with virtual buttons) | 49 (83 with device navigation) |
| Container color | Color: Surface container </br> Elevation: Level 2   |  No content under: #F7F7F7, 80% </br> Content under: Transparent with blur effect |
| Label Typography | Roboto, medium, 12 | San Francisco, medium 11 |
| Icons | Library: [Material Symbols](https://fonts.google.com/icons) </br> Size: 24X24| Library: [SF Symbols](https://developer.apple.com/sf-symbols/) </br> Size: varies, most frequently 25x25. See [documentation](https://developer.apple.com/design/human-interface-guidelines/components/navigation-and-search/tab-bars/) for details. |
| Layout | The exact padding and margins can be found in the [ M3 navigation bar design specs](https://m3.material.io/components/top-app-bar/specs).|  The exact padding and margins can be found in the Sketch design files from [here](https://developer.apple.com/design/resources/#ios-apps).|
| Tab item click animation | Material design 3 uses an animated selection indicator ![mat=animation](https://user-images.githubusercontent.com/21065911/231520113-df1142fb-aa77-41ef-8869-deee841081d8.gif)| HIGs doesn't use an animation ![hig-animation](https://user-images.githubusercontent.com/21065911/231520296-0dfa9dbb-5352-4e38-bfe8-2ad1581ebb16.gif)|

The Flutter framework has two different widgets for displaying bottom navigation bars:

- [NavigationBar](https://api.flutter.dev/flutter/material/NavigationBar-class.html), which implements the specification defined by Material Design 3.
- [BottomNavigationBar](https://api.flutter.dev/flutter/material/BottomNavigationBar-class.html), which implements the specification defined by older versions Material Design.
- [CupertinoTabBar](https://api.flutter.dev/flutter/cupertino/CupertinoTabBar-class.html), which follows the look & feel of the iOS default tab bars.

# Adaptation recommendation
If you're using the default Material styling for your navigation bar on Android, we recommend that the following properties also feel native on iOS.

- Animations
- Font styling (if using system fonts throughout your application)
- Container size
- Container styling

## Rationale and exception
Matching iOS styling can help your Flutter app feel more natural to iOS users. However, many developers chose to implement more custom tab bars, that feature customized animations and / or custom icons. For example, the Twitter mobile app:

![Apr-12-2023 12-34-12](https://user-images.githubusercontent.com/21065911/231523436-2bb16987-3c8d-4076-9fdb-364046b842af.gif)

Developers can find customized navigation bas on pub.dev, for example in the [custom_navigation_bar_package](https://pub.dev/packages/custom_navigation_bar). 

# Example code
All the properties listed above are not currently able to be modified directly from Material's `NavigationBar` widget. Therefore, we recommend adapting to platform-specific navigation bars. 

To implement platform-specific navigation bars, you can use Flutter’s `NavigationBar` widget on Android and the `CupertinoTabBar` widget on iOS. Below is a code snippet you can adapt to show a platform-specific navigation bars. 

```dart
// Create a common object to store the items. This item should have a string (label) and a corresponding Icon.
// In this example, we adapt the Icon based on the platform.
// This common object is helpful since CupertinoTabBar and NavigationBar require different data types.
final Map<String, Icon> _navigationItems = {
    'Menu': Platform.isIOS ? Icon(CupertinoIcons.house_fill) : Icon(Icons.home),
    'Order': Icon(Icons.adaptive.share),
  };

...

Scaffold(
        body: _currentWidget,
        bottomNavigationBar: Platform.isIOS
                ? CupertinoTabBar(
                    currentIndex: _currentIndex,
                    onTap: (index) {
                      setState(() => _currentIndex = index);
                      _loadScreen();
                    },
                    // Map the common entries to the BottomNavigationBarItem type required by the CupertinoTabBar.
                    items: _navigationItems.entries
                        .map<BottomNavigationBarItem>(
                            (entry) => BottomNavigationBarItem(
                                  icon: entry.value,
                                  label: entry.key,
                                ))
                        .toList(),
                  )
                : NavigationBar(
                    selectedIndex: _currentIndex,
                    onDestinationSelected: (index) {
                      setState(() => _currentIndex = index);
                      _loadScreen();
                    },
                    // Map the common entries to the NavigationDestination type required by the NavigationBar.
                    destinations: _navigationItems.entries
                        .map<Widget>((entry) => NavigationDestination(
                              icon: entry.value,
                              label: entry.key,
                            ))
                        .toList(),
                  ));
```

**Comment by bernaferrari** [comment]
I haven't used iOS for a while, so correct me if I'm wrong, but isn't their AppBar more often than not (like on App Store app) translucent to the content behind while Material usually is not?

**Reply by leighajarett** [reply]
This is a bottom tab bar, AppBar is discussed in more detail here: https://github.com/flutter/uxr/discussions/93

**Reply by bernaferrari** [reply]
Yeah, but the AppStore is translucent, at least on iPad.

**Reply by bernaferrari** [reply]
Here:
<img width="598" alt="image" src="https://user-images.githubusercontent.com/351125/231542670-5b76e9d3-eb2a-4cc0-8ac3-b55d8897b239.png">

**Reply by leighajarett** [reply]
Oh you mean tab bar, yes I tried to make that more clear in the container color section - thank you!

**Reply by bernaferrari** [reply]
sorry, I messed up with the names lol

**Reply by leighajarett** [reply]
Hahaha all good! Do you find this information helpful? Anything you would want added or more clearly explained?

**Reply by bernaferrari** [reply]
Yeah, it is more clear now. I don't understand the height "108 with device navigation", maybe could be more clearer or include examples.

**Reply by bernaferrari** [reply]
Oh... I understand it, but what if the user has 3 button navigation? Then it is not 108, lol...

**Reply by leighajarett** [reply]
The height is consistent regardless of the number of buttons :)

**Reply by bernaferrari** [reply]
No, the OS height, isn't 108 the bar + the OS navigation bar?

**Reply by leighajarett** [reply]
108 = tab bar height (80) + os navigation bar for Material on Android

**Reply by bernaferrari** [reply]
yes, but the OS navigation bar changes. I don't use the "line" OS bar in my Android, so my bar is taller.

**Reply by bernaferrari** [reply]
![Screenshot_20230412_161649_Google Play Store.png](https://user-images.githubusercontent.com/351125/231561912-2e8144fc-1b32-4d86-acb5-3d04a348232d.png)

**Reply by leighajarett** [reply]
Ahh I understand, that's why we provide both numbers. So you can take the actual height of the bar (80) and add it to what your os nav bar is. Maybe we can make that more clear!

**Reply by bernaferrari** [reply]
Yeah, but there are two different possible bars, 108 is one of them, mine would be larger.

**Reply by bernaferrari** [reply]
This one is smaller:
<img width="304" alt="image" src="https://user-images.githubusercontent.com/351125/231562344-0a2fd753-4717-4a8f-ab05-17aa7ce4d6ee.png">

**Reply by leighajarett** [reply]
Right so you would take the 80 and add it to whatever your os navigation bar is.

**Reply by bernaferrari** [reply]
Yeah, just wanted to mention "it is not always fixed". I've seen people in iOS using fixed values (and it works fine there). But on Android they don't.

**Reply by leighajarett** [reply]
I tried to add that in to make it more clear, what do you think? Would you reword?

**Reply by bernaferrari** [reply]
It is better. I would probably do:
80 (108 with pill-shaped device navigation, higher on devices with virtual buttons)
But either way is fine. I just wanted you to know because on iOS you can really hard-code 83 and not worry. On Android you can never do this.
Thanks!

**Comment by InMatrix** [comment]
Hi @leighajarett, I tried the suggestion in this guide to adapt Veggie Seasons' bottom navigation bar. While it's not hard to make it work, I found a few issues. 

First, it's difficult to share `_navigationItems.entries` between  `CupertinoTabBar.items` and `NavigationBar.destinations`. The technical issue here is that these two parameters expect different types of objects as their respective value. And the UX issue is that the icon styles are quite different between iOS and Material, so I need to switch the icons too. Here's [the code](https://github.com/InMatrix/veggieseasons_adaptive/commit/498e584ca74846eb7344cf9864a26ad7dd6d994f#diff-3c4164a8346a87854665f200ffb80f341d7b09f063ca86e40d8d6f6dd1109509R75-R126) I ended up with. 

Second, a very minor issue  – the recommendation above the code snippet is to use `CupertinoNavigationBar` on iOS, which is a different widget from `CupertinoTabBar`. 

Last, the initial recommendation gives me the impression that I should try to customize the Material `NavigationBar`, but then the code example is about switching to `CupertinoTabBar`. Maybe you wanted to describe two separate options depending on how much brand-driven customization is done in the app?

**Reply by leighajarett** [reply]
Thanks Tao! 

1. Yes I am aware of this issue, that is why I mapped the data to the proper type in the code sample. I also did adapt the icons - did you see the first object? 

2. I updated, thanks for the catch!

3. I was trying to give the impression that the listed properties should feel native, but since we don't have access to edit those properties directly in the Material widget I swapped to Cupertino. I can try and make that more clear!

**Reply by leighajarett** [reply]
I have added some inline comments to make 1 more clear, and rephrased to address 3. If you have other suggestions please let me know!

**Reply by InMatrix** [reply]
Thank you for your clarification. I've updated my code to use the shared item list. Here's the commit: https://github.com/InMatrix/veggieseasons_adaptive/commit/1f5dc2a48ee96219251e5eddc25d1361df5a4fec.
